scrapy/core/engine.py

```diff
@@ -257,17 +257,15 @@ def _schedule_request(self, request: Request, spider: Spider) -> None:
 
     def download(self, request: Request, spider: Optional[Spider] = None) -> Deferred:
         """Return a Deferred which fires with a Response as result, only downloader middlewares are applied"""
-        if spider is None:
-            spider = self.spider
-        else:
+        if spider is not None:
             warnings.warn(
                 "Passing a 'spider' argument to ExecutionEngine.download is deprecated",
                 category=ScrapyDeprecationWarning,
                 stacklevel=2,
             )
             if spider is not self.spider:
                 logger.warning("The spider '%s' does not match the open spider", spider.name)
-        if spider is None:
+        if self.spider is None:
```

Comment:
Hm, it looks like the behavior is changed in this PR. Previously, if `spider` is passed to `download` method, it is always used. Now a RuntimeError could be raised if `spider` is passed, and `self.spider` is None. Is it a problem?

Reply:
I am unsure, but I will look into that when I find some time to work on tests.

Reply:
For what i understood from PR:

- #5090 

This will be never the case it seems, since the spider instance saved `ExecutionEngine.spider` after calling `open_spider` will be always available, did i get correctly?

Reply:
@kmike @Gallaecio Lookin in all the tests we always pass the spider instance in `open_spider` (except `tests/test_pipeline_files`, `tests/test_pipeline_media.py`), the test that we can add should be the one to test `RuntimeError` exception then?